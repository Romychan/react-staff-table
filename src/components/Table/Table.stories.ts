import type { Meta, StoryObj } from '@storybook/react';

import {
  MOCK_COLUMNS,
  MOCK_DATA_TABLE,
  MOCK_SORTING,
} from '@tests/mock/tableMockData';

import { Users } from '@interfaces/Users';

import { Table } from './Table';

/** This is a component for displaying information in table. */
const meta: Meta<typeof Table<Users, keyof Users>> = {
  title: 'Components/Table',
  component: Table,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    columns: MOCK_COLUMNS,
    sorting: MOCK_SORTING,
  },
  argTypes: { handleSorting: { action: 'handleSorting' } },
};

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to default. */
export const Default: Story = {
  args: {
    data: MOCK_DATA_TABLE,
    isLoading: false,
  },
};

/** This state is used if data is being loaded. */
export const Loading: Story = {
  args: {
    data: [],
    isLoading: true,
  },
};

/** This state is used if the data array is empty. */
export const Empty: Story = {
  args: {
    data: [],
    isLoading: false,
  },
};
