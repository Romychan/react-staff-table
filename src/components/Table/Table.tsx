import {
  TableOrderValue,
  IColumnTable,
  ISortingTable,
  MinTableItem,
} from '@interfaces/Table';

import { TableHeader } from '@ui/TableHeader/TableHeader';
import { TableBody } from '@ui/TableBody/TableBody';

import styles from './Table.module.scss';

interface ITableProps<T, K extends keyof T> {
  columns: IColumnTable<T, K>[];
  data: T[];
  isLoading?: boolean;
  sorting: ISortingTable<T, K>;
  handleSorting: (accessor: K, order: TableOrderValue) => void;
}

export const Table = <T extends MinTableItem, K extends keyof T>({
  columns,
  data,
  isLoading = false,
  sorting,
  handleSorting,
}: ITableProps<T, K>) => {
  return (
    <table className={styles.table}>
      <TableHeader
        columns={columns}
        sorting={sorting}
        handleSorting={handleSorting}
      />
      <TableBody columns={columns} data={data} isLoading={isLoading} />
    </table>
  );
};
