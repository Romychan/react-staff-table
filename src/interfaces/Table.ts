import { ReactNode } from 'react';

export type TableOrderValue = 'asc' | 'desc';

export interface MinTableItem {
  id: string | number;
}

export interface ISortingTable<T, K extends keyof T> {
  column: K;
  order: TableOrderValue;
}

export interface IRenderedCell<T> {
  currentData: T;
}

export interface IColumnTable<T, K extends keyof T> {
  label: string;
  accessor: K;
  sortable?: boolean;
  order?: string;
  Cell?: (props: IRenderedCell<T>) => ReactNode;
  width?: number;
}
