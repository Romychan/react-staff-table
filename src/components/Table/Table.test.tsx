import { fireEvent, screen, render } from '@testing-library/react';
import { vi } from 'vitest';

import {
  MOCK_COLUMNS,
  MOCK_DATA_TABLE,
  MOCK_SORTING,
} from '@tests/mock/tableMockData';

import { Table } from './Table';

describe('Table', () => {
  it('should render correctly', () => {
    const handleClick = vi.fn();

    render(
      <Table
        columns={MOCK_COLUMNS}
        data={MOCK_DATA_TABLE}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );
    const headerCellText = screen.getByText(/Profile/i);
    const bodyCellText = screen.getByText(/test@test.test/i);

    const headerCellElement = screen.getAllByTestId('table-header-cell')[0];

    expect(headerCellText).toBeInTheDocument();
    expect(bodyCellText).toBeInTheDocument();

    expect(headerCellElement.getAttribute('class')).toMatch(/asc/gi);
  });

  it('should render loading element', () => {
    const handleClick = vi.fn();

    render(
      <Table
        columns={MOCK_COLUMNS}
        data={[]}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
        isLoading
      />,
    );

    const loaderElement = screen.getByTestId('loader');

    expect(loaderElement).toBeInTheDocument();
  });

  it('should render with empty data', () => {
    const handleClick = vi.fn();

    render(
      <Table
        columns={MOCK_COLUMNS}
        data={[]}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );

    expect(screen.getByText(/Empty Data/i)).toBeInTheDocument();
  });

  it('should handle click if sortable', () => {
    const handleClick = vi.fn();

    render(
      <Table
        columns={MOCK_COLUMNS}
        data={MOCK_DATA_TABLE}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );

    const cellHeaderElement = screen.getAllByTestId('table-header-cell')[0];

    expect(cellHeaderElement.getAttribute('class')).toMatch(/asc/gi);

    fireEvent.click(cellHeaderElement);
    expect(handleClick).toHaveBeenCalledTimes(1);

    expect(cellHeaderElement.getAttribute('class')).toMatch(/desc/gi);
  });

  it('should not handle click if not sortable', () => {
    const handleClick = vi.fn();

    render(
      <Table
        columns={MOCK_COLUMNS}
        data={MOCK_DATA_TABLE}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );
    const cellHeaderElement = screen.getAllByTestId('table-header-cell')[2];

    fireEvent.click(cellHeaderElement);
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  it('should match snapshot', () => {
    const handleClick = vi.fn();

    const container = render(
      <Table
        columns={MOCK_COLUMNS}
        data={MOCK_DATA_TABLE}
        sorting={MOCK_SORTING}
        handleSorting={handleClick}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
