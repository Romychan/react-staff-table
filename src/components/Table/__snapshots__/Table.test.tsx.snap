// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="_table_6f0cba"
      >
        <thead
          class="_header_692233"
        >
          <tr
            class="_row_692233"
          >
            <th
              class="_cell_d2877c _sortable_d2877c _asc_d2877c"
              data-testid="table-header-cell"
            >
              <div
                class="_container_d2877c"
              >
                <svg
                  class="icon icon-arrow _icon_d2877c"
                  data-testid="icon"
                  height="15"
                  stroke=""
                  width="15"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#arrow"
                  />
                </svg>
                ID
              </div>
            </th>
            <th
              class="_cell_d2877c _sortable_d2877c"
              data-testid="table-header-cell"
            >
              <div
                class="_container_d2877c"
              >
                <svg
                  class="icon icon-arrow _icon_d2877c"
                  data-testid="icon"
                  height="15"
                  stroke=""
                  width="15"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#arrow"
                  />
                </svg>
                Profile
              </div>
            </th>
            <th
              class="_cell_d2877c"
              data-testid="table-header-cell"
            >
              <div
                class="_container_d2877c"
              >
                Email
              </div>
            </th>
          </tr>
        </thead>
        <tbody
          class="_body_2d6dd5"
          data-testid="table-body"
        >
          <tr
            class="_row_2d6dd5"
          >
            <td
              class="_cell_2d6dd5"
            >
              1
            </td>
            <td
              class="_cell_2d6dd5"
            >
              <button>
                Gilberte Prin
              </button>
            </td>
            <td
              class="_cell_2d6dd5"
            >
              test@test.test
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="_table_6f0cba"
    >
      <thead
        class="_header_692233"
      >
        <tr
          class="_row_692233"
        >
          <th
            class="_cell_d2877c _sortable_d2877c _asc_d2877c"
            data-testid="table-header-cell"
          >
            <div
              class="_container_d2877c"
            >
              <svg
                class="icon icon-arrow _icon_d2877c"
                data-testid="icon"
                height="15"
                stroke=""
                width="15"
              >
                <use
                  xlink:href="/src/assets/img/general/icons.svg#arrow"
                />
              </svg>
              ID
            </div>
          </th>
          <th
            class="_cell_d2877c _sortable_d2877c"
            data-testid="table-header-cell"
          >
            <div
              class="_container_d2877c"
            >
              <svg
                class="icon icon-arrow _icon_d2877c"
                data-testid="icon"
                height="15"
                stroke=""
                width="15"
              >
                <use
                  xlink:href="/src/assets/img/general/icons.svg#arrow"
                />
              </svg>
              Profile
            </div>
          </th>
          <th
            class="_cell_d2877c"
            data-testid="table-header-cell"
          >
            <div
              class="_container_d2877c"
            >
              Email
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        class="_body_2d6dd5"
        data-testid="table-body"
      >
        <tr
          class="_row_2d6dd5"
        >
          <td
            class="_cell_2d6dd5"
          >
            1
          </td>
          <td
            class="_cell_2d6dd5"
          >
            <button>
              Gilberte Prin
            </button>
          </td>
          <td
            class="_cell_2d6dd5"
          >
            test@test.test
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
